Pyautogui: A PyAutoGUI é a "porta de entrada" mais famosa e simples para
automatizar o nosso computador pessoal. A PyAutoGUI é uma biblioteca
de automação de interface gráfica(GUI) para Python. O objetivo dela
é simples: fazer o que um humano faz com o mouse e o teclado.

-> Ela é considerada "agnóstica", o que significa que não importa se você
está tentando clicar em um botão de um site no Chrome, em uma celula do
Excel ou em um botão dentro de um jogo; se o elemento está na tela, a
PyAutoGUI pode interagir com ele.

Os 3 Pilares da PyAutoGUI:

1. Controle do Mouse: Ela permite mover o cursor para qualquer coordenada
(x,y) da sua tela, clicar (botão direito, esquerdo ou meio), dar dois cliques
e até arrastar janelas.

2. Controle do Teclado: Ela simula a digitação de textos ou pressionamento de
teclas especiais (como enter, tab, f1, ou combinações como ctrl + c)

3. Funções de "Visão" Básica: Ela consegue tirar screenshots e procurar se
uma imagem especifica (que você salvou antes) aparece em algum lugar da tela
para retornar as coordenadas dela.

In [ ]:
# Import da bibliotecsa PyAutoGUI que possui as funções necessárias para 
# as automações de tarefa 
import pyautogui

# Import da biblioteca time que tem como objetivo manipular as funções que lidam com o tempo.
# Vamos utilizar ela para dar intervalos de segundos na execução do código.
import time

                    INTRODUÇÃO AO PYAUTOGUI: PRINCIPAIS FUNÇÕES

O position tem como objetivo retornar a atual posição do cursor do mouse

In [24]:
print(pyautogui.position())

Point(x=760, y=886)


x: quantidade de pixel para a direita (distância do inicio da tela)

y: Quantidade de pixels para baixo

Ira mostrar o tamanho da sua tela (resolução da tela)

In [ ]:
print(pyautogui.size())

Size(width=1920, height=1080)


Temos 1920 pixels de largura e 1080 pixels de altura

O click tem como objetivo clicar nos elementos da tela

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código      
time.sleep(5)

# A função click recebe como argumento as coordenadas  do local onde
# você quer que o cursor do mouse clique. 
# Observação: As coordenadas podem ser adquiridas usando a função position.
# Basta posicionar o mouse no elemento desejado e executar a função.
pyautogui.click(x=1550, y=186)

moveTo: Tem como objetivo mover o cursor do mouse para o elemento desejado

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código
time.sleep(5)

# A função moveTO recebe como argumento as coordenadas x e y (mesma lógica
# do clique) e o duration que define o quão rápido o cursor irá se mover.
pyautogui.moveTo(x=1787, y=147, duration=1)

PAUSE: Como tempo de carregamento de uma página pode causar falhas na nossa
automação, devemos usar o atributo PAUSE que define o tempo que o sistema deve
demorar para realizar a próxima ação. Dessa maneira, temos tempo suficiente
para executar todas as ações da automação, pois, o site tem tempo de carregar
completamente a página.

In [ ]:
# O atributo recebe como valor o tempo que o intervalo deve durar.
pyautogui.PAUSE = 0.3 

# Após o intervalo, ira mover o cursor para as coordenadas desejadas
pyautogui.moveTo(x=1868, y=178, duration=1)

# Ira clicar no elemento presente nas coordenadas especificadas
pyautogui.click(x=1868, y=178)

# Após o intervalo, ira mover o cursor para as coordenadas desejadas
pyautogui.moveTo(x=1033, y=272, duration=1)

# Ira clicar no elemento presente nas coordenadas especificadas
pyautogui.click(x=1033, y=272)

    Scroll: Tem como objetivo "rolar" a tela para baixo ou para cima

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código.
time.sleep(5)

# O scroll recebe como argumento um valor numérico que define o quanto a tela
# deve ser rolada.

# Valores negativos: irá rolar a tela para baixo
# valores positivos: Irá rolar a tela para cima
pyautogui.scroll(-100)

Write: Tem como objetivo escrever textos no local que mouse esta posicionado

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código.
time.sleep(5)

# A função recebe como argumento o texto que deve ser escrito no local
# desejado
pyautogui.write("Preencha o formulario")

Resultado da função write: Preencha o formulario

Hotkey: Tem como objetivo pressionar teclas do nosso teclado

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código
time.sleep(5)

# O hotckey recebe como argumento a tecla (ou o conjunto de teclas) que deve ser
# pressionada. Observação: Vale ressaltar que geralmente utilizamos o hotkey
# para realizar combinações, pois, diferente do press (que funciona apenas
# para uma única tecla) o hotkey recebe mais de um argumento
pyautogui.hotkey("ctrl", "v")

Resultado do ctrl + v que solicitamos usando o hotkey: Preencha o formulario

Keyboard_keys: Atributo que possui uma lista com os nomes das teclas que queremos pressionar usando o hotkey ou o press.

In [ ]:
print(pyautogui.KEYBOARD_KEYS)

['\t', '\n', '\r', ' ', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~', 'accept', 'add', 'alt', 'altleft', 'altright', 'apps', 'backspace', 'browserback', 'browserfavorites', 'browserforward', 'browserhome', 'browserrefresh', 'browsersearch', 'browserstop', 'capslock', 'clear', 'convert', 'ctrl', 'ctrlleft', 'ctrlright', 'decimal', 'del', 'delete', 'divide', 'down', 'end', 'enter', 'esc', 'escape', 'execute', 'f1', 'f10', 'f11', 'f12', 'f13', 'f14', 'f15', 'f16', 'f17', 'f18', 'f19', 'f2', 'f20', 'f21', 'f22', 'f23', 'f24', 'f3', 'f4', 'f5', 'f6', 'f7', 'f8', 'f9', 'final', 'fn', 'hanguel', 'hangul', 'hanja', 'help', 'home', 'insert', 'junja', 'kana', 'kanji', 'launchapp1', 'launchapp2', 'launchmai

Press: Temo como objetivo pressionar teclas do nosso teclado (uma tecla por vez, pois ele não aceita mais de um argumento).

In [ ]:
# Ira dar um intervalo de 5 segundos na execução do código
time.sleep(5)

# A função press recebe como argumento o nome da tecla (presente no atributo keyword_keys)
# que queremos pressionar. 
# Observação: podemos adicionar o argumento "presses" que define quantas vezes
# queremos que o sistema pressione a tecla especificada.
pyautogui.press("win")

                            Automatizando um login
-> Agora vamos realizar um exercicio que tem como objetivo realizar o preenchimento de um formulário de login de forma automatica no site da
hashtag treinamentos.

Fluxo de passos da automação:

-> Mover o cursor até o botão de login

-> Clicar no botão de login

-> Clicar no campo de email

-> Escrever o email no campo clicado

-> Mover o cursor até o botão de envio do formulário

-> Clicar no botão de envio de formulário

In [ ]:
# Ira dar o intervalo de 5 segundos que possibilita o nosso acesso
# a página que será controlada pela automação do PyAutoGUI
time.sleep(5)

# Ira mover o cursor ate o botão de login com duração de 1 segundo
# (movimentos suaves)
pyautogui.moveTo(x=1342, y=162, duration=1)

# Ira clicar no botão de login
pyautogui.click(x=1342, y=162)

# Ira dar uma pausa de 3 segundos com o objetivo de só prosseguir
# com a execução do código após o carregamento completo da página.
pyautogui.PAUSE = 3

# Após o carregamento da página, iremos mover o cursor até o campo
# de email da tela de login.
pyautogui.moveTo(x=1093, y=490, duration=1)

# Ira clicar no campo de email do login
pyautogui.click(x=1093, y=490)

# Ira dar uma pausa de 1 segundo
pyautogui.PAUSE = 1

# Após o clique, iremos escrever o email no campo
pyautogui.write("email@gmail.com")

# Após escrever, iremos mover o cursor para o botão de envio do formulário
pyautogui.moveTo(x=1103, y=619, duration=1)

# Por ultimo iremos clicar no botão de envio de formulário
pyautogui.click(x=1103, y=619)


